knowledge_repo/app/auth_providers/oauth2.py

```diff
@@ -0,0 +1,165 @@
+import json
+from flask import request, render_template, flash, redirect, url_for
+from flask_login import login_user
+from six.moves.urllib.parse import urljoin
+
+from ..models import User
+from ..proxies import db_session
+from ..utils.auth import is_safe_url
+
+from ..auth_provider import KnowledgeAuthProvider
+
+
+PRESETS = {
+    'bitbucket': {
+        'base_url': 'https://api.bitbucket.org/2.0/',
+        'authorization_url': 'https://bitbucket.org/site/oauth2/authorize',
+        'token_url': 'https://bitbucket.org/site/oauth2/access_token',
+        'auto_refresh_url': 'https://bitbucket.org/site/oauth2/access_token',
+        'scopes': 'account',
+        'user_info_endpoint': 'user',
+        'user_info_mapping': {
+            'identifier': 'username',
+            'name': 'name',
+            'photo_uri': ['links', 'avatar']
+        }
+    },
+    'github': {
+        'base_url': 'https://api.github.com/',
+        'authorization_url': 'https://github.com/login/oauth/authorize',
+        'token_url': 'https://github.com/login/oauth/access_token',
+        'auto_refresh_url': None,
+        'scopes': None,
+        'user_info_endpoint': 'user',
+        'user_info_mapping': {
+            'identifier': 'email',
+            'name': 'name',
+            'photo_uri': 'avatar_url'
+        }
+    },
+    'google': {
+        'base_url': 'https://www.googleapis.com/oauth2/v2/',
+        'authorization_url': 'https://accounts.google.com/o/oauth2/auth',
+        'token_url': 'https://accounts.google.com/o/oauth2/token',
+        'auto_refresh_url': 'https://accounts.google.com/o/oauth2/token',
+        'scopes': [
+            'https://www.googleapis.com/auth/userinfo.email',
+            'https://www.googleapis.com/auth/userinfo.profile'
+        ],
+        'user_info_endpoint': 'userinfo',
+        'user_info_mapping': {
+            'identifier': 'email',
+            'name': 'name',
+            'photo_uri': 'picture'
+        }
+    }
+}
+
+
+def _resolve_oauth_config(name, local_namespace, config, *variables):
+    presets = PRESETS.get(name, {})
+    output = []
+    for variable in variables:
+        value = local_namespace.get(variable, None)
+        if value is None:
```

Comment:
In this case, there is a difference between things like `[]`, `False`, `''` and `None`, which may vary between OAuth providers. If any "False-like" object is passed that is not `None`, it should be the value passed on.